database/migrations/upcoming/2023_05_12_000000_utf8mb4.php

```diff
@@ -0,0 +1,44 @@
+<?php
+
+use Illuminate\Database\Migrations\Migration;
+
+return new class() extends Migration {
+    public function up(): void
+    {
+        $this->migrateCharsetTo('utf8mb4', 'utf8mb4_unicode_ci');
+    }
+
+    /**
+     * Reverse the migrations.
+     */
+    public function down(): void
+    {
+        // do not revert
+        // $this->migrateCharsetTo('latin1', 'latin1_general_ci');
+    }
+
+    protected function migrateCharsetTo(string $charset, string $collation): void
+    {
+        $defaultConnection = config('database.default');
+        $databaseName = config("database.connections.{$defaultConnection}.database");
+
+        DB::unprepared("ALTER SCHEMA `{$databaseName}` DEFAULT CHARACTER SET {$charset} DEFAULT COLLATE {$collation};");
+
+        $tableNames = DB::table('information_schema.tables')
+            ->where('table_schema', $databaseName)
+            ->pluck('TABLE_NAME');
+
+        $ignoredTables = [
+            'Activity',
+            'Awarded',
+        ];
+
+        foreach ($tableNames as $tableName) {
+            if (in_array($tableName, $ignoredTables)) {
+                continue;
+            }
+            echo PHP_EOL . '   Converting ' . $tableName;
+            DB::unprepared("ALTER TABLE {$tableName} CONVERT TO CHARACTER SET {$charset} COLLATE {$collation};");
```

Comment:
I parked this migration in `upcoming` after coming to the same conclusion - to make the conversion complete a data migration has to happen right afterwards to convert utf8 characters. The plan was to switch to utf8mb on the db connection level with v3 but this would lead to a lot of errors when mixing latin tables with utf8mb4 connections.

See https://github.com/RetroAchievements/RAWeb/blob/v3/config/database.php#L57-L61

This means that the new migrations will all create new tables and columns with latin charset - which doesn't make much of a difference now if we plan on migrating the whole db at once in the future.